File: android/src/main/java/com/capacitor/cors/PluginInterceptor.java
package com.capacitor.cors;

import android.util.Log;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import org.json.JSONException;
import org.json.JSONObject;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

/**
 * Plugin Interceptor Manager
 * Manages custom interceptors for HTTP requests
 */
public class PluginInterceptor implements Interceptor {
    private static final String TAG = "PluginInterceptor";
    
    private final List<InterceptorEntry> interceptors = Collections.synchronizedList(new ArrayList<>());
    private final AtomicInteger interceptorCounter = new AtomicInteger(0);
    private final CorsBypassPlugin plugin;

    public PluginInterceptor(CorsBypassPlugin plugin) {
        this.plugin = plugin;
    }

    /**
     * Interceptor Entry
     */
    public static class InterceptorEntry {
        public String id;
        public String name;
        public int priority;
        public boolean enabled;
        public InterceptorScope scope;
        
        public InterceptorEntry(String id, String name, int priority, boolean enabled, InterceptorScope scope) {
            this.id = id;
            this.name = name;
            this.priority = priority;
            this.enabled = enabled;
            this.scope = scope;
        }
    }

    /**
     * Interceptor Scope
     */
    public static class InterceptorScope {
        public String urlPattern;
        public List<String> methods;
        
        public InterceptorScope(String urlPattern, List<String> methods) {
            this.urlPattern = urlPattern;
            this.methods = methods;
        }
        
        public boolean matches(String url, String method) {
            // Check URL pattern
            if (urlPattern != null && !url.matches(urlPattern)) {
                return false;
            }
            
            // Check method
            if (methods != null && !methods.isEmpty() && !methods.contains(method)) {
                return false;
            }
            
            return true;
        }
    }

    /**
     * Add interceptor
     */
    public String addInterceptor(JSObject options) {
        String id = "interceptor_" + interceptorCounter.incrementAndGet();
        String name = options.optString("name", null);
        int priority = options.optInt("priority", 0);
        boolean enabled = options.optBoolean("enabled", true);
        
        // Parse scope
        InterceptorScope scope = null;
        if (options.has("scope")) {
            try {
                Object scopeValue = options.get("scope");
                if (scopeValue instanceof org.json.JSONObject) {
                    org.json.JSONObject scopeJson = (org.json.JSONObject) scopeValue;
                    String urlPattern = scopeJson.has("urlPattern") ? scopeJson.getString("urlPattern") : null;
                    List<String> methods = null;
                    
                    if (scopeJson.has("methods")) {
                        org.json.JSONArray methodsArray = scopeJson.getJSONArray("methods");
                        if (methodsArray != null) {
                            methods = new ArrayList<>();
                            for (int i = 0; i < methodsArray.length(); i++) {
                                methods.add(methodsArray.getString(i));
                            }
                        }
                    }
                    
                    scope = new InterceptorScope(urlPattern, methods);
                }
            } catch (Exception e) {
                Log.e(TAG, "Error parsing scope", e);
            }
        }
        
        InterceptorEntry entry = new InterceptorEntry(id, name, priority, enabled, scope);
        interceptors.add(entry);
        
        // Sort by priority (higher first)
        Collections.sort(interceptors, new Comparator<InterceptorEntry>() {
            @Override
            public int compare(InterceptorEntry a, InterceptorEntry b) {
                return Integer.compare(b.priority, a.priority);
            }
        });
        
        Log.d(TAG, "Added interceptor: " + id + " (priority: " + priority + ")");
        return id;
    }

    /**
     * Remove interceptor
     */
    public boolean removeInterceptor(String id) {
        synchronized (interceptors) {
            for (int i = 0; i < interceptors.size(); i++) {
                if (interceptors.get(i).id.equals(id)) {
                    interceptors.remove(i);
                    Log.d(TAG, "Removed interceptor: " + id);
                    return true;
                }
            }
        }
        return false;
    }

    /**
     * Remove all interceptors
     */
    public void removeAllInterceptors() {
        synchronized (interceptors) {
            interceptors.clear();
            Log.d(TAG, "Removed all interceptors");
        }
    }

    /**
     * Get all interceptors
     */
    public JSArray getAllInterceptors() {
        JSArray result = new JSArray();
        synchronized (interceptors) {
            for (InterceptorEntry entry : interceptors) {
                JSObject obj = new JSObject();
                obj.put("id", entry.id);
                if (entry.name != null) {
                    obj.put("name", entry.name);
                }
                obj.put("enabled", entry.enabled);
                result.put(obj);
            }
        }
        return result;
    }

    /**
     * Enable/disable interceptor
     */
    public boolean setInterceptorEnabled(String id, boolean enabled) {
        synchronized (interceptors) {
            for (InterceptorEntry entry : interceptors) {
                if (entry.id.equals(id)) {
                    entry.enabled = enabled;
                    Log.d(TAG, (enabled ? "Enabled" : "Disabled") + " interceptor: " + id);
                    return true;
                }
            }
        }
        return false;
    }

    /**
     * OkHttp Interceptor implementation
     */
    @Override
    public Response intercept(Chain chain) throws IOException {
        Request originalRequest = chain.request();
        
        // Create request context
        JSObject requestConfig = new JSObject();
        requestConfig.put("url", originalRequest.url().toString());
        requestConfig.put("method", originalRequest.method());
        
        // Convert headers
        JSObject headers = new JSObject();
        for (String name : originalRequest.headers().names()) {
            headers.put(name, originalRequest.header(name));
        }
        requestConfig.put("headers", headers);
        
        // Execute request interceptors
        JSObject modifiedConfig = executeRequestInterceptors(requestConfig);
        
        // Build modified request if needed
        Request request = originalRequest;
        if (modifiedConfig != null && !modifiedConfig.toString().equals(requestConfig.toString())) {
            Request.Builder builder = originalRequest.newBuilder();
            
            // Update headers
            if (modifiedConfig.has("headers")) {
                try {
                    Object headersValue = modifiedConfig.get("headers");
                    if (headersValue instanceof org.json.JSONObject) {
                        org.json.JSONObject modifiedHeaders = (org.json.JSONObject) headersValue;
                        // Clear existing headers and add modified ones
                        Iterator<String> keys = modifiedHeaders.keys();
                        while (keys.hasNext()) {
                            String key = keys.next();
                            builder.header(key, modifiedHeaders.getString(key));
                        }
                    }
                } catch (Exception e) {
                    Log.e(TAG, "Error updating headers", e);
                }
            }
            
            request = builder.build();
        }
        
        // Execute request
        Response response;
        try {
            response = chain.proceed(request);
            
            // Execute response interceptors
            JSObject responseData = new JSObject();
            responseData.put("status", response.code());
            responseData.put("statusText", response.message());
            responseData.put("url", response.request().url().toString());
            
            // Convert response headers
            JSObject responseHeaders = new JSObject();
            for (String name : response.headers().names()) {
                responseHeaders.put(name, response.header(name));
            }
            responseData.put("headers", responseHeaders);
            
            JSObject modifiedResponse = executeResponseInterceptors(responseData);
            
            // Note: We can't easily modify the response body in OkHttp interceptor
            // So we just log the interceptor execution
            if (modifiedResponse != null) {
                Log.d(TAG, "Response interceptors executed");
            }
            
            return response;
            
        } catch (IOException e) {
            // Execute error interceptors
            JSObject error = new JSObject();
            error.put("message", e.getMessage());
            error.put("config", requestConfig);
            
            JSObject handledResponse = executeErrorInterceptors(error);
            
            if (handledResponse != null) {
                Log.d(TAG, "Error handled by interceptor");
                // Note: Can't easily create a Response from JSObject in interceptor
                // This would need to be handled at a higher level
            }
            
            throw e;
        }
    }

    /**
     * Execute request interceptors
     */
    private JSObject executeRequestInterceptors(JSObject config) {
        JSObject modifiedConfig = config;
        
        synchronized (interceptors) {
            for (InterceptorEntry entry : interceptors) {
                if (!entry.enabled) {
                    continue;
                }
                
                // Check scope
                if (entry.scope != null) {
                    String url = config.optString("url", "");
                    String method = config.optString("method", "");
                    if (!entry.scope.matches(url, method)) {
                        continue;
                    }
                }
                
                // Notify plugin to execute JS interceptor
                JSObject event = new JSObject();
                event.put("type", "onRequest");
                event.put("interceptorId", entry.id);
                event.put("config", modifiedConfig);
                
                // Note: This is a simplified version
                // In a real implementation, we would need to wait for JS response
                Log.d(TAG, "Request interceptor: " + entry.id);
            }
        }
        
        return modifiedConfig;
    }

    /**
     * Execute response interceptors
     */
    private JSObject executeResponseInterceptors(JSObject response) {
        JSObject modifiedResponse = response;
        
        synchronized (interceptors) {
            for (InterceptorEntry entry : interceptors) {
                if (!entry.enabled) {
                    continue;
                }
                
                // Notify plugin to execute JS interceptor
                JSObject event = new JSObject();
                event.put("type", "onResponse");
                event.put("interceptorId", entry.id);
                event.put("response", modifiedResponse);
                
                Log.d(TAG, "Response interceptor: " + entry.id);
            }
        }
        
        return modifiedResponse;
    }

    /**
     * Execute error interceptors
     */
    private JSObject executeErrorInterceptors(JSObject error) {
        synchronized (interceptors) {
            for (InterceptorEntry entry : interceptors) {
                if (!entry.enabled) {
                    continue;
                }
                
                // Notify plugin to execute JS interceptor
                JSObject event = new JSObject();
                event.put("type", "onError");
                event.put("interceptorId", entry.id);
                event.put("error", error);
                
                Log.d(TAG, "Error interceptor: " + entry.id);
            }
        }
        
        return null;
    }
}

File: ios/Plugin/PluginInterceptor.swift
import Foundation
import Capacitor

/**
 * Plugin Interceptor Manager
 * Manages custom interceptors for HTTP requests
 */
@objc public class PluginInterceptor: NSObject {
    
    // MARK: - Types
    
    struct InterceptorEntry {
        let id: String
        var name: String?
        var priority: Int
        var enabled: Bool
        var scope: InterceptorScope?
    }
    
    struct InterceptorScope {
        let urlPattern: String?
        let methods: [String]?
        
        func matches(url: String, method: String) -> Bool {
            // Check URL pattern
            if let pattern = urlPattern {
                let regex = try? NSRegularExpression(pattern: pattern, options: [])
                let range = NSRange(url.startIndex..., in: url)
                if regex?.firstMatch(in: url, options: [], range: range) == nil {
                    return false
                }
            }
            
            // Check method
            if let methods = methods, !methods.isEmpty, !methods.contains(method) {
                return false
            }
            
            return true
        }
    }
    
    // MARK: - Properties
    
    private var interceptors: [InterceptorEntry] = []
    private var interceptorCounter = 0
    private let lock = NSLock()
    private weak var plugin: CorsBypassPlugin?
    
    // MARK: - Initialization
    
    init(plugin: CorsBypassPlugin) {
        self.plugin = plugin
        super.init()
    }
    
    // MARK: - Public Methods
    
    /**
     * Add interceptor
     */
    func addInterceptor(options: [String: Any]) -> String {
        lock.lock()
        defer { lock.unlock() }
        
        interceptorCounter += 1
        let id = "interceptor_\(interceptorCounter)"
        let name = options["name"] as? String
        let priority = options["priority"] as? Int ?? 0
        let enabled = options["enabled"] as? Bool ?? true
        
        // Parse scope
        var scope: InterceptorScope?
        if let scopeDict = options["scope"] as? [String: Any] {
            let urlPattern = scopeDict["urlPattern"] as? String
            let methods = scopeDict["methods"] as? [String]
            scope = InterceptorScope(urlPattern: urlPattern, methods: methods)
        }
        
        let entry = InterceptorEntry(
            id: id,
            name: name,
            priority: priority,
            enabled: enabled,
            scope: scope
        )
        
        interceptors.append(entry)
        
        // Sort by priority (higher first)
        interceptors.sort { $0.priority > $1.priority }
        
        print("[PluginInterceptor] Added interceptor: \(id) (priority: \(priority))")
        return id
    }
    
    /**
     * Remove interceptor
     */
    func removeInterceptor(id: String) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        
        if let index = interceptors.firstIndex(where: { $0.id == id }) {
            interceptors.remove(at: index)
            print("[PluginInterceptor] Removed interceptor: \(id)")
            return true
        }
        return false
    }
    
    /**
     * Remove all interceptors
     */
    func removeAllInterceptors() {
        lock.lock()
        defer { lock.unlock() }
        
        interceptors.removeAll()
        print("[PluginInterceptor] Removed all interceptors")
    }
    
    /**
     * Get all interceptors
     */
    func getAllInterceptors() -> [[String: Any]] {
        lock.lock()
        defer { lock.unlock() }
        
        return interceptors.map { entry in
            var dict: [String: Any] = [
                "id": entry.id,
                "enabled": entry.enabled
            ]
            if let name = entry.name {
                dict["name"] = name
            }
            return dict
        }
    }
    
    /**
     * Enable/disable interceptor
     */
    func setInterceptorEnabled(id: String, enabled: Bool) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        
        if let index = interceptors.firstIndex(where: { $0.id == id }) {
            interceptors[index].enabled = enabled
            print("[PluginInterceptor] \(enabled ? "Enabled" : "Disabled") interceptor: \(id)")
            return true
        }
        return false
    }
    
    // MARK: - Interceptor Execution
    
    /**
     * Execute request interceptors
     */
    func executeRequestInterceptors(config: [String: Any]) -> [String: Any] {
        lock.lock()
        let activeInterceptors = interceptors.filter { $0.enabled }
        lock.unlock()
        
        var modifiedConfig = config
        
        for entry in activeInterceptors {
            // Check scope
            if let scope = entry.scope {
                let url = config["url"] as? String ?? ""
                let method = config["method"] as? String ?? ""
                if !scope.matches(url: url, method: method) {
                    continue
                }
            }
            
            // Notify plugin to execute JS interceptor
            let event: [String: Any] = [
                "type": "onRequest",
                "interceptorId": entry.id,
                "config": modifiedConfig
            ]
            
            print("[PluginInterceptor] Request interceptor: \(entry.id)")
            
            // Note: In a real implementation, we would need to wait for JS response
            // For now, we just log the execution
        }
        
        return modifiedConfig
    }
    
    /**
     * Execute response interceptors
     */
    func executeResponseInterceptors(response: [String: Any]) -> [String: Any] {
        lock.lock()
        let activeInterceptors = interceptors.filter { $0.enabled }
        lock.unlock()
        
        var modifiedResponse = response
        
        for entry in activeInterceptors {
            // Notify plugin to execute JS interceptor
            let event: [String: Any] = [
                "type": "onResponse",
                "interceptorId": entry.id,
                "response": modifiedResponse
            ]
            
            print("[PluginInterceptor] Response interceptor: \(entry.id)")
        }
        
        return modifiedResponse
    }
    
    /**
     * Execute error interceptors
     */
    func executeErrorInterceptors(error: [String: Any]) -> [String: Any]? {
        lock.lock()
        let activeInterceptors = interceptors.filter { $0.enabled }
        lock.unlock()
        
        for entry in activeInterceptors {
            // Notify plugin to execute JS interceptor
            let event: [String: Any] = [
                "type": "onError",
                "interceptorId": entry.id,
                "error": error
            ]
            
            print("[PluginInterceptor] Error interceptor: \(entry.id)")
        }
        
        return nil
    }
}

/**
 * URLProtocol-based interceptor for URLSession
 */
class InterceptorURLProtocol: URLProtocol {
    
    private static var interceptor: PluginInterceptor?
    
    static func setInterceptor(_ interceptor: PluginInterceptor) {
        self.interceptor = interceptor
    }
    
    override class func canInit(with request: URLRequest) -> Bool {
        // Check if we've already handled this request
        if URLProtocol.property(forKey: "HandledByInterceptor", in: request) != nil {
            return false
        }
        
        // Only intercept if we have an interceptor configured
        return interceptor != nil
    }
    
    override class func canonicalRequest(for request: URLRequest) -> URLRequest {
        return request
    }
    
    override func startLoading() {
        guard let interceptor = InterceptorURLProtocol.interceptor else {
            client?.urlProtocolDidFinishLoading(self)
            return
        }
        
        // Execute request interceptors
        var config: [String: Any] = [
            "url": request.url?.absoluteString ?? "",
            "method": request.httpMethod ?? "GET"
        ]
        
        if let headers = request.allHTTPHeaderFields {
            config["headers"] = headers
        }
        
        let modifiedConfig = interceptor.executeRequestInterceptors(config: config)
        
        // Create modified request if needed
        let modifiedRequest = (request as NSURLRequest).mutableCopy() as! NSMutableURLRequest
        if let modifiedHeaders = modifiedConfig["headers"] as? [String: String] {
            for (key, value) in modifiedHeaders {
                modifiedRequest.setValue(value, forHTTPHeaderField: key)
            }
        }
        
        // Mark the request as handled to prevent infinite loops
        URLProtocol.setProperty(true, forKey: "HandledByInterceptor", in: modifiedRequest)
        
        // Execute the actual request
        let task = URLSession.shared.dataTask(with: modifiedRequest as URLRequest) { [weak self] data, response, error in
            guard let self = self else { return }
            
            if let error = error {
                // Execute error interceptors
                let errorDict: [String: Any] = [
                    "message": error.localizedDescription,
                    "config": config
                ]
                _ = interceptor.executeErrorInterceptors(error: errorDict)
                
                self.client?.urlProtocol(self, didFailWithError: error)
                return
            }
            
            if let response = response as? HTTPURLResponse {
                // Execute response interceptors
                var responseDict: [String: Any] = [
                    "status": response.statusCode,
                    "statusText": HTTPURLResponse.localizedString(forStatusCode: response.statusCode),
                    "url": response.url?.absoluteString ?? ""
                ]
                
                var headers: [String: String] = [:]
                for (key, value) in response.allHeaderFields {
                    if let keyString = key as? String, let valueString = value as? String {
                        headers[keyString] = valueString
                    }
                }
                responseDict["headers"] = headers
                
                _ = interceptor.executeResponseInterceptors(response: responseDict)
                
                self.client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
            }
            
            if let data = data {
                self.client?.urlProtocol(self, didLoad: data)
            }
            
            self.client?.urlProtocolDidFinishLoading(self)
        }
        
        task.resume()
    }
    
    override func stopLoading() {
        // Cancel any ongoing requests
    }
}